components/encryption/src/manager/mod.rs

```diff
@@ -244,31 +287,34 @@ impl Dicts {
         Ok(())
     }
 
-    fn rotate_key(&mut self, key_id: u64, key: DataKey, master_key: &dyn Backend) -> Result<bool> {
+    fn rotate_key(&self, key_id: u64, key: DataKey, master_key: &dyn Backend) -> Result<bool> {
         info!("encryption: rotate data key."; "key_id" => key_id);
-        match self.key_dict.keys.entry(key_id) {
-            // key id collides
-            Entry::Occupied(_) => return Ok(false),
-            Entry::Vacant(e) => e.insert(key),
+        {
+            let mut key_dict = self.key_dict.lock().unwrap();
+            match key_dict.keys.entry(key_id) {
+                // key id collides
+                Entry::Occupied(_) => return Ok(false),
+                Entry::Vacant(e) => e.insert(key),
+            };
         };
-        // Update current data key id.
-        self.key_dict.current_key_id = key_id;
         // re-encrypt key dict file.
-        self.save_key_dict(master_key).map(|()| true)
+        let ret = self.save_key_dict(master_key).map(|()| true);
+        // Update current data key id.
+        self.current_key_id.store(key_id, Ordering::SeqCst);
```

Comment:
update only when ret.ok()

Reply:
I think `save_key_dict` should be executed after update.

Reply:
If we update before save and if the save failed (IO error, etc), a subsequent `new_file` call can get the un-saved `current_key_id` and use it to encrypt a new file, and we will lose the key for decrypt after restart.

Reply:
Here you can set key_dict.current_key_id to the new current key id, and update `self.current_key_id` after save success.